src/ddsrt/src/library/posix/library.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright(c) 2006 to 2018 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <stdio.h>
+#include <dlfcn.h>
+#include <assert.h>
+#include <string.h>
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/library.h"
+
+
+dds_retcode_t
+ddsrt_dlopen(
+    const char *name,
+    bool translate,
+    ddsrt_lib *handle)
+{
+  dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+  if ( name == NULL || handle == NULL ){
+      retcode = DDS_RETCODE_BAD_PARAMETER;
+
+  } else {
+      *handle = NULL;
+      dlerror();    /* Clear any existing error */
+      if ( (translate) && (strrchr(name, '/') == NULL)) {
+        /* Add lib and suffix to the name and try to open. */
+#if __APPLE__
+        static const char suffix[] = ".dylib";
+#else
+        static const char suffix[] = ".so";
+#endif
+        char* libName = ddsrt_malloc(3 + strlen(name) + sizeof(suffix));
+        sprintf(libName, "lib%s%s", name, suffix);
+        *handle = dlopen(libName, RTLD_GLOBAL | RTLD_NOW);
+        ddsrt_free(libName);
+
+
+      }
+
+      if (*handle == NULL && retcode == DDS_RETCODE_OK) {
+        /* name contains a path,
+         * (auto)translate is disabled or
+         * dlopen on translated name failed. */
+        *handle = dlopen(name, RTLD_GLOBAL | RTLD_NOW);
+
+      }
+
+      if ( *handle != NULL ){
+          retcode = DDS_RETCODE_OK;
+      } else {
+          retcode = DDS_RETCODE_ERROR;
+      }
+  }
+
+  return retcode;
+}
+
+
+dds_retcode_t
+ddsrt_dlclose(
+    ddsrt_lib handle)
+{
+    dds_retcode_t retcode;
+
+    if( handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    }
+    else{
+        retcode = (dlclose(handle) == 0) ? DDS_RETCODE_OK : DDS_RETCODE_ERROR;
+    }
+
+    return retcode;
+}
+
+dds_retcode_t
+ddsrt_dlsym(
+    ddsrt_lib handle,
+    const char *symbol,
+    void **address)
+{
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+    if ( symbol == NULL || handle == NULL || address == NULL){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    } else {
+        *address =  dlsym(handle, symbol);
```

Comment:
Use a local variable and only assign `address` on success.

Reply:
I think it is safer to reset the given argument to NULL if there is an error and the dlsym already sets it to NULL in case of error.

Reply:
I'm have a bigger problem with the coding style (this is not the only place) of declaring-and-initialising a return value to "OK" and then changing it on error. *If* one insists on single return and initialising all variables at declaration time as "safety" measures, *then* the logical thing would be to initialise a return code to "ERROR".

Reply:
The initial result value is OK because, in this approach, there can be different error codes; the correct one will be assigned according to conditions. If there had only one error code and several valid (OK) return codes, I would have used ERROR initialisation at the beginning.